PopPUNK/network.py

```diff
@@ -704,3 +724,112 @@ def printExternalClusters(newClusters, extClusterFile, outPrefix,
                                 index = False)
+
+def generate_minimum_spanning_tree(G, names, distMat):
+    """Generate a minimum spanning tree from a network
+
+    Args:
+       G (network)
+           Graph tool network
+       names (list)
+           List of sequence names
+
+    Returns:
+       tree_string (str)
+           Newick representation of the minimum spanning tree
+    """
+    sys.stderr.write("Starting calculation of minimum-spanning tree\n")
+    # Define sequences names for tree
+    taxon_namespace = dendropy.TaxonNamespace(names)
+    # Test if weighted network
+    weighted_network = False
+    if "weight" in G.edge_properties:
+        weighted_network = True
+    # Calculate minimum spanning tree
+    if weighted_network:
+        mst_edge_prop_map = gt.min_spanning_tree(G, weights = G.ep["weight"])
+    else:
+        mst_edge_prop_map = gt.min_spanning_tree(G)
+    mst_network = gt.GraphView(G, efilt = mst_edge_prop_map)
+    # Find seed nodes as those with greatest outdegree in each component
+    seed_vertices = set()
+    component_assignments, component_frequencies = gt.label_components(mst_network)
+    for component_index in range(len(component_frequencies)):
+        component_members = component_assignments.a == component_index
+        component = gt.GraphView(mst_network, vfilt = component_members)
+        component_vertices = component.get_vertices()
+        out_degrees = component.get_out_degrees(component_vertices)
+        seed_vertex = list(component_vertices[np.where(out_degrees == np.amax(out_degrees))])
+        seed_vertices.add(seed_vertex[0]) # Can only add one otherwise not MST
+    # If multiple components, calculate distances between seed nodes
+    if len(component_frequencies) > 1:
+        # Extract distances
+        connections = []
+        for row_idx, (ref, query) in enumerate(listDistInts(names, names, self = True)):
+            if ref in seed_vertices and query in seed_vertices:
+                dist = distMat[row_idx, 0]
+                edge_tuple = (ref, query, dist)
+                connections.append(edge_tuple)
+        # Construct graph
+        seed_G = gt.Graph(directed = False)
+        seed_G.add_vertex(len(seed_vertex))
+        if weighted_network:
+            eweight = seed_G.new_ep("float")
+            seed_G.add_edge_list(connections, eprops = [eweight])
+            seed_G.edge_properties["weight"] = eweight
+            seed_mst_edge_prop_map = gt.min_spanning_tree(seed_G, weights = seed_G.ep["weight"])
+            seed_mst_network = gt.GraphView(seed_G, efilt = seed_mst_edge_prop_map)
+            # Insert seed MST into original MST - may be possible to use graph_union with include=True & intersection
+            deep_edges = seed_mst_network.get_edges([seed_mst_network.ep["weight"]])
+            mst_network.add_edge_list(deep_edges)
+        else:
+            sys.stderr.write('I do not currently know what to do in this case\n')
```

Comment:
Please make this more informative, e.g. 'unweighted minimum spanning tree not implemented'

Reply:
This was just a placeholder to remind myself to put some code here. I think it requires extracting the sparse matrix concatenation code from the lineage models.

Reply:
I am removing the ability to do this with unweighted graphs

Reply:
Sensible.